core/src/main/scala/cats/data/HashSet.scala

```diff
@@ -0,0 +1,937 @@
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package cats.data
+
+import cats.Show
+import cats.UnorderedFoldable
+import cats.kernel.CommutativeMonoid
+import cats.kernel.Hash
+import cats.kernel.instances.StaticMethods
+import cats.syntax.eq._
+import java.util.Arrays
+
+/**
+  * An immutable hash set using [[cats.kernel.Hash]] for hashing.
+  *
+  * Implemented using the CHAMP encoding.
+  *
+  * @see [[https://michael.steindorfer.name/publications/phd-thesis-efficient-immutable-collections.pdf Efficient Immutable Collections]]
+  *
+  * @tparam A the type of the elements contained in this hash set.
+  * @param hash the [[cats.kernel.Hash]] instance used for hashing values.
+  */
+final class HashSet[A](
+  private val rootNode: HashSet.Node[A]
+)(implicit hash: Hash[A]) {
+
+  /**
+    * An iterator for this set that can be used only once.
+    *
+    * @return an iterator that iterates through the elements of this set.
+    */
+  final def iterator: Iterator[A] =
+    new HashSet.Iterator(rootNode)
+
+  /**
+    * A reverse iterator for this set that can be used only once.
+    *
+    * @return an iterator that iterates through this set in the reverse order of [[HashSet#iterator]].
+    */
+  final def reverseIterator: Iterator[A] =
+    new HashSet.ReverseIterator(rootNode)
+
+  /**
+    * The size of this set.
+    *
+    * @return the number of elements in this set.
+    */
+  final def size: Int = rootNode.size
+
+  /**
+    * Tests whether the set is empty.
+    *
+    * @return `true` if the set contains no elements, `false` otherwise.
+    */
+  final def isEmpty: Boolean = size == 0
+
+  /**
+    * Tests whether the set is not empty.
+    *
+    * @return `true` if the set contains at least one element, `false` otherwise.
+    */
+  final def nonEmpty: Boolean = !isEmpty
+
+  /**
+    * Apply `f` to each element for its side effects.
+    *
+    * @param f the function to apply to each element.
+    */
+  final def foreach[U](f: A => U): Unit =
+    rootNode.foreach(f)
+
+  /**
+    * Test whether the set contains `value`.
+    *
+    * @param value the element to check for set membership.
+    * @return `true` if the set contains `value`, `false` otherwise.
+    */
+  final def contains(value: A): Boolean =
+    rootNode.contains(value, hash.hash(value), 0)
```

Comment:
Some hash tries don't use the hash code directly, but "improve" it (in case the hash function is not very good). There is `scala.collection.Hashing.improve` which is not public, but there is also `scala.util.hashing.byteswap32`, which _is_ public.

Reply:
This is interesting: I guess my question is, how do you "improve" an existing hash without adding any new information to it? If the existing hash function isn't very good, won't the results of "improve" be just as bad since it is a pure function? Or is it about how the existing information in the hash is distributed across the bits of the hash?

I can't find a lot of information about the `improve` function in the Scala collections - it looks like it has been around since Scala 2.3!

https://github.com/scala/scala/commit/fee21b7e701bbb4a3f311cf06ecd668cb5d0bc94#diff-86c85e543a3eed2597cca1172039d2036015f7d471e106e94f3f0e9785009b42R58=

Reply:
Ah there is more explanation of the `improve` function here: https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/mutable/HashTable.scala#L403=
It looks like `byteswap32` is preferred over the previous implementation

Reply:
So I guess my next question is, should we assume that the input `hash` is not very well distributed if it is produced by the `Hash` instance for that type?

Reply:
Resolved in https://github.com/typelevel/cats/pull/4185/commits/b497ab100e8f1b384ab1faf1da0a9be5ea016936

Reply:
Cool. Yes, I think it's worth assuming that `Hash` might not be very good. It's certainly true for `Int` and `Long`. And even if the hash is good, I think it's unlikely that running it through `byteswap32` would make it _worse_.